src/logic/map_objects/tribes/ship.cc

```diff
@@ -608,6 +620,82 @@ bool Ship::ship_update_transport(Game& game, Bob::State& state) {
 	return true;
 }
 
+void Ship::send_known_portspace_message (Game& game) {
+	const std::string& icon_filename =
+	   ship_type_ == ShipType::kWarship ? descr().icon_filename() : kPortspaceIconFile;
+	const std::string message_body = ship_type_ == ShipType::kWarship ?
+                                    _("A warship arrived at a known port build space.") :
+                                    _("An expedition ship arrived at a known port build space.");
+	send_message(game, _("Port Space"), _("Port Space Spotted"), message_body, icon_filename);
+}
+void Ship::send_new_portspace_message (Game& game) {
+	const std::string& icon_filename =
+	   ship_type_ == ShipType::kWarship ? descr().icon_filename() : kPortspaceIconFile;
+	const std::string message_body = ship_type_ == ShipType::kWarship ?
+                                   _("A warship found a new port build space.") :
+                                   _("An expedition ship found a new port build space.");
+	send_message(game, _("Port Space"), _("Port Space Found"), message_body, icon_filename);
+}
+
+bool Ship::update_seen_portspaces(Game& game) {
+	if (expedition_ == nullptr) {
+		return false;
+	}
+
+	bool lost_sight = false;
+	bool spotted = false;
+	const Map& map = owner().egbase().map();
+
+	const std::function<bool(const Coords&)> is_suitable = [this](const Coords& coords) {
```

Comment:
This should be a member function of Ship

Reply:
OK. I almost made it like that, but this was supposed to be a minor optimisation. But thinking more about it, calling this is already filtered for actual port spaces, so it isn't called all that often after all.